pkg/securitypolicy/internal.go

```diff
@@ -0,0 +1,254 @@
+)
+
+type mountInternal struct {
+	Source      string
+	Destination string
+	Type        string
+	Options     []string
+}
+
+// newMountInternal creates an internal mount constraint object from given
+// source, destination, type and options
+func newMountInternal(src, dst string, mType string, mOpts []string) mountInternal {
+	return mountInternal{
+		Source:      src,
+		Destination: dst,
+		Type:        mType,
+		Options:     mOpts,
+	}
+}
+
+// securityPolicyContainer is an internal version of Container
+type securityPolicyContainer struct {
+	// The command that we will allow the container to execute
+	Command []string
+	// The rules for determining if a given environment variable is allowed
+	EnvRules []EnvVarRule
+	// An ordered list of dm-verity root hashes for each layer that makes up
+	// "a container". Containers are constructed as an overlay file system. The
+	// order that the layers are overlayed is important and needs to be enforced
+	// as part of policy.
+	Layers []string
+	// WorkingDir is a path to container's working directory, which all the processes
+	// will default to.
+	WorkingDir string
+	// Unordered list of mounts which are expected to be present when the container
+	// starts
+	ExpectedMounts []string `json:"expected_mounts"`
+	// A list of constraints for determining if a given mount is allowed.
+	Mounts []mountInternal
+}
+
+// validate checks given OCI mount against mount policy. Source and destination
+// are checked by direct string comparisons. Different matching strategies can
+// be added by introducing a separate path matching config, which isn't needed
+// at the moment.
+func (m *mountInternal) validate(mSpec oci.Mount) error {
+	if m.Type != mSpec.Type {
+		return fmt.Errorf("mount type not allowed by policy: expected=%q, actual=%q", m.Type, mSpec.Type)
+	}
+	if ok, _ := regexp.MatchString(m.Source, mSpec.Source); !ok {
+		return fmt.Errorf("mount source not allowed by policy: expected=%q, actual=%q", m.Source, mSpec.Source)
+	}
+	if m.Destination != mSpec.Destination && m.Destination != "" {
+		return fmt.Errorf("mount destination not allowed by policy: expected=%q, actual=%q", m.Destination, mSpec.Destination)
+	}
+	if !cmp.Equal(m.Options, mSpec.Options) {
+		return fmt.Errorf("mount options not allowed by policy: expected=%q, actual=%q", m.Options, mSpec.Options)
+	}
+	return nil
+}
+
+// matchMount matches given OCI mount against mount constraints. If no match
+// found, the mount is not allowed.
+func (c *securityPolicyContainer) matchMount(sandboxID string, m oci.Mount) (err error) {
+	mountOk := false
+	for _, constraint := range c.Mounts {
+		// now that we know the sandboxID we can get the actual path for
+		// various destination path types by adding a UVM mount prefix
+		logrus.Debugf("validating constraint: %+v", constraint)
+		constraint = substituteUVMPath(sandboxID, constraint)
+		if err = constraint.validate(m); err == nil {
+			mountOk = true
+			break
+		}
+	}
+	if !mountOk {
```

Comment:
We either set `mountOk` and break from the loop, in which case we will always return nil, or we end the loop without setting `mountOk` and thus hit this conditional. Can we simplify by just doing `return nil` from inside the loop, and always return an error if we get to down here?